src/query/stream/streams.cpp

```diff
@@ -190,26 +191,29 @@ Streams::StreamsMap::iterator Streams::CreateConsumer(StreamsMap &map, const std
 
   auto *memory_resource = utils::NewDeleteResource();
 
-  auto consumer_function =
-      [interpreter_context = interpreter_context_, memory_resource, stream_name,
-       transformation_name = stream_info.common_info.transformation_name, owner = owner,
-       interpreter = std::make_shared<Interpreter>(interpreter_context_),
-       result = mgp_result{nullptr, memory_resource}](const std::vector<typename TStream::Message> &messages) mutable {
-        auto accessor = interpreter_context->db->Access();
-        EventCounter::IncrementCounter(EventCounter::MessagesConsumed, messages.size());
-        CallCustomTransformation(transformation_name, messages, result, accessor, *memory_resource, stream_name);
-
-        DiscardValueResultStream stream;
-
-        spdlog::trace("Start transaction in stream '{}'", stream_name);
-        utils::OnScopeExit cleanup{[&interpreter, &result]() {
-          result.rows.clear();
-          interpreter->Abort();
-        }};
+  auto consumer_function = [interpreter_context = interpreter_context_, memory_resource, stream_name,
+                            transformation_name = stream_info.common_info.transformation_name, owner = owner,
+                            interpreter = std::make_shared<Interpreter>(interpreter_context_),
+                            result = mgp_result{nullptr, memory_resource},
+                            total_retries = interpreter_context_->config.stream_transaction_conflict_retries](
+                               const std::vector<typename TStream::Message> &messages) mutable {
+    auto accessor = interpreter_context->db->Access();
+    EventCounter::IncrementCounter(EventCounter::MessagesConsumed, messages.size());
+    CallCustomTransformation(transformation_name, messages, result, accessor, *memory_resource, stream_name);
+
+    DiscardValueResultStream stream;
+
+    spdlog::trace("Start transaction in stream '{}'", stream_name);
+    utils::OnScopeExit cleanup{[&interpreter, &result]() {
+      result.rows.clear();
+      interpreter->Abort();
+    }};
+
+    const static std::map<std::string, storage::PropertyValue> empty_parameters{};
+    const bool shall_retry = total_retries != 0;
+    for (uint32_t i = 1; i != total_retries && shall_retry; ++i) {
```

Comment:
```suggestion
    for (uint32_t i = 0; i <= total_retries; ++i) {
```
I am a bit confused. I think the current implementation doesn't run the query at all if the retry count is `0` or `1`. [Godbolted](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYgATKVpMGoZAFI9AIQuXSS%2BsgJ4BlRugDCqWgFcWDEJIAHKRuADJ4DJgAcr4ARpjEEgDspAAOqAqEzgyePn4BwemZTgLhkTEs8YlcKfaYjtlCBEzEBLm%2B/kF2mA4lDE0tBGXRcQnJds2t7fldCpNDESOVYzUAlHao3sTI7BxmGgCCFgDMEcg%2BWADUZsfuyAjEAqg32PtHeqcM595XN%2B5U3i%2BfTELzeJzOF0w11uzjmxEwrFBh3BX0h0PccIiwCR70%2B31%2BtwIDwR6BxKPxUL%2BgJIWHh6AA%2BiwmKkyR8IT9KbdvE5aIQAJ5kw4RAiXdCoABKmAIxD5EDQDDml2FlyIzVo9Ph0rwmAUq2uSWsh0uxsuc3QIBQmxFfz%2B1z0egGrUuAHdCAgVag1RqpcRtQoQHa9OjbaqxN6tTrg7dLmA9gBWdwMWM3Q0HE2mgjmkCYkyXVLwmiqaEAEUD4sBDCxKoQUIAjt4EnzSEqAxY9Cm3un5YrYqgvKaEGJ1ZqZSWPV6R36Y2AbqWNB2jSb%2BMRLhBvMLjnp6SK8GP58dLErp7Px2HJ5GLAA2K8DofhmUpu3WKx4PVmA2d9Mm6V8/Wpr8ARmWZoNyUbuHmBZ4EWNrRruMHgcmCZJnsB6foBJr5pghZjm2ko/lWRJ1g2MrNhQgYLmm6Hfg8qDOkBFrEICThsPSCSPMQEBtkwCikvaqwUeh76lqIBD3Ku3YimaDFMXgLFsSQN6YG%2BH6LlRSpUKucHHMWJ6hsOPp%2Bsp/5qV%2BUmWqB8GBgAYgcACSoTYMWYAzkhbYCSZ6aalsDCXAAtFw7lqUJaFUd4mS5swbAKKkTA7PRKA0QwqD0ryBAJGICiBVRZlEngCjbsSTBZgo9CYKk9LLhAXAaCwupZemwWqfqOlNWZIHWrctptkI3jIDsCgKM5ZiufaWVecQPn7qmjVHEKgiihKPp8pYjDaKgcoCIqyp6fehl/iFbVWmBgaOiKrpEqe%2BkRv6gbHTt54KMdiGJsmqGtZmFo5sAEFYVBOH2uWDCVrmhGXPWjaka2o1vZRxrLqu66CJu25HieU2o512mXbtOqPjYNivvtTXpj%2BROw9lH3mR17i2ph2GWVpNPRs9yH1YBdN/SeuFLQRNZg8RTYtuRMMeUSjx0WZjGCLJmCscQ7Gcfa3G8Xo/Ei4BQmXCJYkbQqkmU1LzGy/JxCKUZIXoXgGkQFpOlY/dBk6ubxMmYdFmY%2BBbY2fZjlDSN7bqx5xrjT5/lswBM0eWFVaRTqMVxW1iXJal6W0JlgcU1muX5WLJKfaV5WVdVtVq8ZEdJC15ORyabvU119o9X1OqDS5iZuRnIeXFNYIV2Cc0ikyEQQM75PwxATDcqgGNzo%2BjOXJIeNWATI/obXx1uR8Jxb5vHxC22x2M7aLOvWXX4SZcE9EJc8IKN4tAdaWYp4TKNulxbNeU%2B16/2k/S3X1K3lMDmlupZG%2Bd867Rjcq5f2WVq7GjXpZKBiYAACqC0HoIwaghaz9lqrWeDAjOY9L5T1tl3We6J56L2fJYQm75T7pgQR7ciO9jjb1YbvMi%2B8GZPXjC9FC9CTTn2If/W%2B98xy/x/CtBga1X7h3il/RBP9FqSLwf/AggDgFcKYWAsRij2zQLbtDU%2BM0hIcHWLQTgcZeD%2BA4FoUgqBOA0yXlYU0mxtiUg%2BHoXgBBNBmPWAAaxAMcQIAA6JIl5LwaGOJEgAnEEKJ9pDCcEkNY3x9jOC8H9BoUgPjbFmNIHAWASA0AsFSHQBI5BKAlLKfQRI5xjDAC4HoDQXA%2BB0DSsQf0EBYhpNiBEFofJOA8FICUtgggADyDBaCDLyaQLATITDiFmfgeEDQABuOo0mYFUPUbkuw7HCh6Gk3ksRiADM8FgNJWoWBDPyVQIwwAFAADVtTOnGakRgtyZCCBEGIdgUhvnyCUGoNJugDBGBMCgfGVhDB4FiP6SA6xUCpD6P6DgvkzSznMM4ywzS/LjOOJkno9Q%2BiuAYB4LwHQJAhHJcMCoVRqVFCyAIaY/hWlMr6HS0Y1Rui9EaPMVl1K6gNAEKdLlyweVzEGIK1pUrWjioZVwdYCg3E7H0Dk%2BEuweDmMsak2ZDiOCqECJeXyl5JBawhd9JpISNAhK4KuJx1Dmy4EICQO0m5myeFKeUlcJxVbeN8asdYNYipjGHqQQJegkghJiRoDQl5jgaDjEkLgxrjiSDjOC5JvAbnVWyTYuxBrMkgGybkrQ6xCmIBAFaVI3JKkbW9bU9VQDXWJAMPwH5ohxAAo7UClQ6hZnSkwFq3gzozmpFuTqjgVjSAFt4Aa8Z3Ja0ilQBpI1JqzUWoaZca1tr7UQC9TUhI7r/U5MDQEoJxwQnHBvbeu9d6kkcBSbOtJRa7AlrPXkoNj6vEvv1Rkz95b1gbM6dkAIQA). With the proposed changes I think we can eliminate `shall_retry` and fix the bug also.

Am I missing something here?

Reply:
This won't work either because it will run a retry `total_retries + 1`. For example if `total_retries == 1` then the loop will execute for `0` and `1` which is `2` retries. I ended up having a `for-loop`. Also, there was another bug. We did not break the loop after committing the transaction which means we redo the transaction a total of `total_retries`